macros.lisp

```diff
@@ -10,44 +10,134 @@ If TEST is `jtruep' evaluate THEN, otherwise evaluate ELSE."
-Alias for `jor'."
-  (warn 'deprecated :deprecated 'or_ :replacement "NJSON/ALIASES:OR")
-  `(or ,@args))
-
 (defmacro jand (&rest args)
   "JSON-aware version of `cl:and'."
   `(and ,@(loop for arg in args
                 collecting `(jwhen ,arg ,arg))))
 
-(defmacro and_ (&rest args)
-  "JSON-aware version of `cl:and'.
-Alias for `jand'."
-  (warn 'deprecated :deprecated 'and_ :replacement "NJSON/ALIASES:AND")
-  `(and ,@args))
+(defun check-value (expected indices object)
+  "Check that JSON value in FORM at INDICES is `equal' to EXPECTED specification."
+  (let ((result (jget indices object t)))
+    (or (case expected
+          (:true (eq t result))
+          (:false (eq nil result))
+          (array (arrayp result))
+          (null (hash-table-p result))
+          (t (equal result expected)))
+        (error 'value-mismatch
+               :expected expected :actual result
+               :object (jget (subseq indices 0
+                                     ;; FIXME: Not exactly a perfect heuristic?
+                                     (max 0 (- (length indices) 2)))
+                             object t)))))
+
+;; DESTRUCTURING-PATTERN is not a (&rest destructuring-pattern)
+;; because it might be a vector too.
+(defmacro jbind (destructuring-pattern form &body body)
+  "Match the FORM against DESTRUCTURING-PATTERN.
+The pattern might be:
+- A symbol, in which case the current chosen form is bound to it. If
+  the symbol is _ or NIL, simply skip the form.
+- A literal form:
+  - String or number: compare with `equal'.
+  - Keywords :TRUE, :FALSE, :NULL, :UNDEFINED, matching T, NIL, :NULL,
+    and :UNDEFINED respectively.
+- If the pattern is a list, match the string+pattern pairs inside it
+  to the provided JSON object and resolve them recursively.
+- If the pattern is an inline vector, match it against a JSON array
+  with at least as much elements as provided in the vector. Match
+  every form in the vector against the element with the same index.
+
+If the DESTRUCTURING-PATTERN doesn't match the object, underlying
+`jget' will throw errors for the parts that don't match. See `jget'
+documentation for the types of errors it throws.
+
+Example:
+\(\"hello\" hello \"a\" _ \"b\" b
+ \"array\" #(first second third _))
+
+matches a JSON object
+{\"hello\": 3, \"a\": 8, \"b\": 3, \"c\": null, \"array\": [1, 2, 3, 4]}
+
+and binds
+- HELLO to 3
+- B to 3
+- FIRST to 1
+- SECOND to 2
+- THIRD to 3
+
+It also checks that \"a\" key is present in the object and there's a
+fourth element in the nested array."
+  (let ((form-sym (gensym "BIND-FORM"))
+        (bindings (list)))
+    (labels ((parse-pattern (pattern &optional (current-path (list)))
+               (etypecase pattern
+                 ((or (member :true :false :null :undefined) string real)
+                  (push (cons pattern (copy-list current-path))
+                        bindings))
+                 (list (loop for (key subpattern) on pattern by #'cddr
```

Comment:
Nit: You can replace this `loop` by `alex:doplist`.

Reply:
Yes, but Alexandria is not a dependency of core NJSON :P